### PR TITLE
New unified Ethlance smart contract

### PR DESCRIPTION
**Pull request by madis**
This is work-in-progress commit (will be modified w/ fixup-s and the
debugging code removed).

### Summary

Unify the various contracts into single `Ethlance.sol`. Also simplify the logic and data (contract, database and web application) significantly by removing the standard bounty and rolling our own simpler solution.

> NB! This is work in progress

### CODE REVIEW

**Review by madvas — APPROVED** [review]
nice work 👍

**Comment by madis** [comment]
Getting an error when starting the ethlance node server. I'll reach out tomorrow morning if I can't get it solved myself
```
❯ node server/ethlance_server.js

<...output omitted...>

/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:532
(defn js-invoke
^
TypeError: Cannot read property 'getPastEvents' of null
    at Function.cljs$core$IFn$_invoke$arity$variadic (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:532:1)
    at Function.cljs$core$IFn$_invoke$arity$variadic (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs_web3_next/eth.cljs:49:1)
    at /home/madis/code/district0x/ethlance/server/cljs-runtime/district/server/smart_contracts.cljs:247:24
    at /home/madis/code/district0x/ethlance/server/cljs-runtime/district.server.smart_contracts.js:619:3
    at /home/madis/code/district0x/ethlance/server/cljs-runtime/district.server.smart_contracts.js:629:3
    at Object.sval (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:3438:7)
    at Object.cljs$core$ISeqable$_seq$arity$1 (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:3429:1)
    at Object.cljs$core$seq [as seq] (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:1234:7)
    at cljs.core/seq (/home/madis/code/district0x/ethlance/server/cljs-runtime/district/server/smart_contracts.cljs:236:20)
    at Object.sval (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:3438:7)
    at Object.cljs$core$ISeqable$_seq$arity$1 (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:3429:1)
    at Object.cljs$core$seq [as seq] (/home/madis/code/district0x/ethlance/server/cljs-runtime/cljs/core.cljs:1234:7)
    at cljs.core/seq (/home/madis/code/district0x/ethlance/server/cljs-runtime/district/server/smart_contracts.cljs:259:5)
    at switch__37037__auto__ (/home/madis/code/district0x/ethlance/server/cljs-runtime/district/server/smart_contracts.cljs:259:5)
    at Function.district$server$smart_contracts$replay_past_events_in_order_$_state_machine__37038__auto____1 [as cljs$core$IFn$_invoke$arity$1] (/home/madis/code/district0x/ethlance/server/cljs-runtime/district.server.smart_contracts.js:2189:4)

```

**Comment by madvas** [comment]
What line throws this error? Such error happens when you'd be setting up listener for non-existing contract. I don't see in code event listener being set up somewhere.

File: src/tests/contract/ethlance_test.cljs [hunk 1]
```diff
@@ -6,44 +6,112 @@
...
 
+(defn publish-to-channel-callback [&stuff])
+(defn subscribe->chan [contract method options]
+  (let [out (chan)
+        callback (fn [event]
+                   (println "Got an event" event)
+                   (put! out event)
+                   )
+        ]
+    (smart-contracts/subscribe-events contract metohd options [publish-to-channel-callback])
```
Comment (madis) [inline src/tests/contract/ethlance_test.cljs:23]: This line produces the exception (even with empty callback implementation as it is). Removing it gets rid of the exception (but also of any useful work :( )

The exception stack trace:
```
Testing tests.contract.ethlance-test

/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core.cljs:9524
      (throw (js/Error. (str "Doesn't support name: " x))))))
      ^
Error: Doesn't support name: 
    at Object.cljs$core$name [as name] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core.cljs:9524:7)
    at Function.cljs$core$IFn$_invoke$arity$variadic (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs_web3_next/eth.cljs:34:1)
    at Object.district$server$smart_contracts$subscribe_events [as subscribe_events] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/district/server/smart_contracts.cljs:170:3)
    at Object.tests$contract$ethlance_test$subscribe__GT_chan [as subscribe__GT_chan] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests/contract/ethlance_test.cljs:17:3)
    at inst_45431 (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests/contract/ethlance_test.cljs:37:12)
    at switch__35204__auto__ (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests/contract/ethlance_test.cljs:37:12)
    at Function.tests$contract$ethlance_test$state_machine__35205__auto____1 [as cljs$core$IFn$_invoke$arity$1] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests.contract.ethlance_test.js:386:4)
    at Object.fexpr__40437 [as run_state_machine] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core/async/impl/ioc_helpers.cljs:43:3)
    at Object.cljs$core$async$impl$ioc_helpers$run_state_machine_wrapped [as run_state_machine_wrapped] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core/async/impl/ioc_helpers.cljs:45:1)
    at /home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core/async/impl/ioc_helpers.cljs:56:38
    at G__42453 (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs_promises/async.cljs:30:31)
    at Immediate.cljs$core$async$impl$dispatch$process_messages (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core/async/impl/dispatch.cljs:26:7)
    at processImmediate (internal/timers.js:443:21)
    at process.topLevelDomainCallback (domain.js:136:23)
```

File: contracts/zksync/EthlanceStructs.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    uint offeredAmount = offer.value;
+    IERC20 offeredToken = IERC20(offer.token.tokenContract.tokenAddress);
+    uint depositedAmount = offeredToken.balanceOf(initialOwner);
+    uint allowedToTransfer = offeredToken.allowance(initialOwner, ethlance);
+    require(allowedToTransfer >= offeredAmount, "Offer must equal to deposit");
+    offeredToken.transferFrom(initialOwner, ethlance, offeredAmount);
+    require(offeredToken.balanceOf(ethlance) > 0, "Ethlance must have received the token");
+    offeredToken.transfer(job, offeredAmount);
+    emit ZeJDeb("Processing offered value", 99);
+    // offeredValues.push(offer);
```
Comment (madis) [inline contracts/zksync/EthlanceStructs.sol:70]: Haven't found a way to do this correctly :(

File: contracts/zksync/Ethlance.sol [hunk 3]
```diff
@@ -390,6 +390,20 @@ contract Ethlance is ApproveAndCallFallBack, IERC721Receiver, IERC1155Receiver,
...
+    // decoding example: https://ethereum.stackexchange.com/questions/66600/decoding-parameters-with-inline-assembly-in-solidity
+    // require(jobProxyTarget != address(0));
+
+    address creator;
+    EthlanceStructs.TokenValue[] memory offeredValues;
+    JobType jobType;
+    address[] memory invitedArbiters;
+    bytes memory ipfsData;
+
+    // (creator, offeredValues, jobType, invitedArbiters, ipfsData) = abi.decode(_data, (address, EthlanceStructs.TokenValue[], JobType, address[], bytes));
```
Comment (madis) [inline contracts/zksync/Ethlance.sol:404]: This line reverts the transaction

File: src/tests/contract/ethlance_test.cljs [hunk 4]
```diff
@@ -106,14 +106,20 @@
...
-             _ (println "safe-transfer-from-result" safe-transfer-from-result)
+             ; _ (println "call-data" call-data)
+             ; init-result (<! (ethlance/initialize job-proxy-address))
+             ; _ (println "init-result" init-result)
+             ; safe-transfer-from-result (<! (smart-contracts/contract-send :test-nft :safe-transfer-from [employer ethlance-addr token-id call-data] {:from employer}))
+             ; _ (println "safe-transfer-from-result" safe-transfer-from-result)
+             ; debug-call-data (web3-eth/encode-abi (smart-contracts/instance :ethlance) :try-decode [{:jobType 0}])
+             debug-call-data (web3-eth/encode-abi (smart-contracts/instance :ethlance)
+                                                 :try-decode
+                                                 [{:jobType 0}])
```
Comment (madis) [inline src/tests/contract/ethlance_test.cljs:117]: Not sure why that line fails :confused: Fails with
```
    var error = new Error(message);                        
                ^                                                                                                      
Error: invalid bytes value (arg="_data", coderType="bytes", value={"jobType":0})
    at Object.throwError (/home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/errors.js:68:17)
    at CoderDynamicBytes.encode (/home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/abi-coder.js:540:20)
    at /home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/abi-coder.js:605:59                       
    at Array.forEach (<anonymous>)                                                                                     
    at pack (/home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/abi-coder.js:604:12)
    at CoderTuple.encode (/home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/abi-coder.js:764:16)
    at AbiCoder.encode (/home/madis/code/district0x/ethlance/server/node_modules/ethers/utils/abi-coder.js:897:77)     
    at ABICoder.encodeParameters (/home/madis/code/district0x/ethlance/server/node_modules/web3-eth-abi/src/index.js:96:27)
    at /home/madis/code/district0x/ethlance/server/node_modules/web3-eth-contract/src/index.js:426:24                  
    at Array.map (<anonymous>)                                                                                         
    at Object._encodeMethodABI (/home/madis/code/district0x/ethlance/server/node_modules/web3-eth-contract/src/index.js:425:12)
    at Function.cljs$core$IFn$_invoke$arity$variadic (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core.cljs:532:1)
    at Object.argseq__4743__auto__ [as js_invoke] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs/core.cljs:532:1)                      
    at Object.cljs_web3_next$eth$encode_abi [as encode_abi] (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/cljs_web3_next/eth.cljs:25:1)
    at inst_161629 (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests/contract/ethlance_test.cljs:90:6)
    at switch__35976__auto__ (/home/madis/code/district0x/ethlance/server/tests/cljs-runtime/tests/contract/ethlance_test.cljs:90:6)
```

File: contracts/zksync/Ethlance.sol [hunk 5]
```diff
@@ -189,36 +206,34 @@ contract Ethlance is ApproveAndCallFallBack, IERC721Receiver, IERC1155Receiver,
...
+    address newJob = address(new MutableForwarder()); // This becomes the new proxy
+    address payable newJobPayableAddress = payable(address(uint160(newJob)));
+    uint timestamp = block.number;
+    MutableForwarder(newJobPayableAddress).setTarget(jobProxyTarget);
+
+    EthlanceStructs.transferToJob(_creator, address(this), newJobPayableAddress, _offeredValues);
+    // Job(newJobPayableAddress).initialize(this, _creator, _jobType, _offeredValues, _invitedArbiters);
+    EthlanceStructs.TokenValue[] memory toval;
+    address[] memory arbit;
+    Job(newJobPayableAddress).initialize(this, address(this), _jobType, toval, arbit);
```
Comment (madis) [inline contracts/zksync/Ethlance.sol:230]: All simple values go in, the `Job#initialize` implementation is empty. Why does it revert the transaction.

**Pull request merged**